docs/spec.md

```diff
@@ -801,37 +801,38 @@ Afterwards, within each `process_group`:
 #### Constraints
 
 * (C1) `is_unique(replica_groups)`.
-* (C2) `size(replica_groups)` is defined as:
+* (C2) `rank(replica_groups) = 2`.
+* (C3) `1 <= size(replica_groups)`.
+* (C4) `0 <= replica_groups < N` where `N` is defined as:
   * `num_replicas` if `cross_replica` is used.
   * `num_replicas` if `cross_replica_and_partition` is used.
   * `num_processes` if `flattened_ids` is used.
-* (C3) `0 <= replica_groups < size(replica_groups)`.
-* (C4) If `use_global_device_ids = true`, then `channel_id > 0`.
```

Comment:
Why are we changing the spec? The current formulation works for the potentially ragged shape of `replica_groups`, while the new formulation does not (because rank is only defined on regular tensors).

Reply:
I thought rank is always 2 and raggedness only changes  `dim(replica_groups, 1)`?

Reply:
The way how TensorConstant is defined doesn't support raggedness in the first place.

Reply:
Ah makes sense.